File: Frameworks/libatrac9/libatrac9/LibAtrac9/CSharp/LibAtrac9/Utilities/BitReader.cs
using System;
using System.Diagnostics;

namespace LibAtrac9.Utilities
{
    internal class BitReader
    {
        private byte[] Buffer { get; set; }
        private int LengthBits { get; set; }
        public int Position { get; set; }
        private int Remaining => LengthBits - Position;

        public BitReader(byte[] buffer) => SetBuffer(buffer);

        public void SetBuffer(byte[] buffer)
        {
            Buffer = buffer;
            LengthBits = Buffer?.Length * 8 ?? 0;
            Position = 0;
        }

        public int ReadInt(int bitCount)
        {
            int value = PeekInt(bitCount);
            Position += bitCount;
            return value;
        }

        public int ReadSignedInt(int bitCount)
        {
            int value = PeekInt(bitCount);
            Position += bitCount;
            return Bit.SignExtend32(value, bitCount);
        }

        public bool ReadBool() => ReadInt(1) == 1;

        public int ReadOffsetBinary(int bitCount, OffsetBias bias)
        {
            int offset = (1 << (bitCount - 1)) - (int)bias;
            int value = PeekInt(bitCount) - offset;
            Position += bitCount;
            return value;
        }

        public void AlignPosition(int multiple)
        {
            Position = Helpers.GetNextMultiple(Position, multiple);
        }

        public int PeekInt(int bitCount)
        {
            Debug.Assert(bitCount >= 0 && bitCount <= 32);

            if (bitCount > Remaining)
            {
                if (Position >= LengthBits) return 0;

                int extraBits = bitCount - Remaining;
                return PeekIntFallback(Remaining) << extraBits;
            }

            int byteIndex = Position / 8;
            int bitIndex = Position % 8;

            if (bitCount <= 9 && Remaining >= 16)
            {
                int value = Buffer[byteIndex] << 8 | Buffer[byteIndex + 1];
                value &= 0xFFFF >> bitIndex;
                value >>= 16 - bitCount - bitIndex;
                return value;
            }

            if (bitCount <= 17 && Remaining >= 24)
            {
                int value = Buffer[byteIndex] << 16 | Buffer[byteIndex + 1] << 8 | Buffer[byteIndex + 2];
                value &= 0xFFFFFF >> bitIndex;
                value >>= 24 - bitCount - bitIndex;
                return value;
            }

            if (bitCount <= 25 && Remaining >= 32)
            {
                int value = Buffer[byteIndex] << 24 | Buffer[byteIndex + 1] << 16 | Buffer[byteIndex + 2] << 8 | Buffer[byteIndex + 3];
                value &= (int)(0xFFFFFFFF >> bitIndex);
                value >>= 32 - bitCount - bitIndex;
                return value;
            }
            return PeekIntFallback(bitCount);
        }

        private int PeekIntFallback(int bitCount)
        {
            int value = 0;
            int byteIndex = Position / 8;
            int bitIndex = Position % 8;

            while (bitCount > 0)
            {
                if (bitIndex >= 8)
                {
                    bitIndex = 0;
                    byteIndex++;
                }

                int bitsToRead = Math.Min(bitCount, 8 - bitIndex);
                int mask = 0xFF >> bitIndex;
                int currentByte = (mask & Buffer[byteIndex]) >> (8 - bitIndex - bitsToRead);

                value = (value << bitsToRead) | currentByte;
                bitIndex += bitsToRead;
                bitCount -= bitsToRead;
            }
            return value;
        }

        /// <summary>
        /// Specifies the bias of an offset binary value. A positive bias can represent one more
        /// positive value than negative value, and a negative bias can represent one more
        /// negative value than positive value.
        /// </summary>
        /// <remarks>Example:
        /// A 4-bit offset binary value with a positive bias can store
        /// the values 8 through -7 inclusive.
        /// A 4-bit offset binary value with a positive bias can store
        /// the values 7 through -8 inclusive.</remarks>
        public enum OffsetBias
        {
            Negative = 0
        }
    }
}

File: Frameworks/libatrac9/libatrac9/LibAtrac9/C/bit_reader.c
#include "bit_reader.h"
#include "utility.h"

static int peek_int_fallback(bit_reader_cxt* br, int bit_count);

void init_bit_reader_cxt(bit_reader_cxt* br, const void * buffer)
{
	br->buffer = buffer;
	br->position = 0;
}

int read_int(bit_reader_cxt* br, const int bits)
{
	const int value = peek_int(br, bits);
	br->position += bits;
	return value;
}

int read_signed_int(bit_reader_cxt* br, const int bits)
{
	const int value = peek_int(br, bits);
	br->position += bits;
	return SignExtend32(value, bits);
}

int read_offset_binary(bit_reader_cxt* br, const int bits)
{
	const int offset = 1 << (bits - 1);
	const int value = peek_int(br, bits) - offset;
	br->position += bits;
	return value;
}

int peek_int(bit_reader_cxt* br, const int bits)
{
	const int byte_index = br->position / 8;
	const int bit_index = br->position % 8;
	const unsigned char* buffer = br->buffer;

	if (bits <= 9)
	{
		int value = buffer[byte_index] << 8 | buffer[byte_index + 1];
		value &= 0xFFFF >> bit_index;
		value >>= 16 - bits - bit_index;
		return value;
	}

	if (bits <= 17)
	{
		int value = buffer[byte_index] << 16 | buffer[byte_index + 1] << 8 | buffer[byte_index + 2];
		value &= 0xFFFFFF >> bit_index;
		value >>= 24 - bits - bit_index;
		return value;
	}

	if (bits <= 25)
	{
		int value = buffer[byte_index] << 24
			| buffer[byte_index + 1] << 16
			| buffer[byte_index + 2] << 8
			| buffer[byte_index + 3];

		value &= (int)(0xFFFFFFFF >> bit_index);
		value >>= 32 - bits - bit_index;
		return value;
	}
	return peek_int_fallback(br, bits);
}

void align_position(bit_reader_cxt* br, const unsigned int multiple)
{
	const int position = br->position;
	if (position % multiple == 0)
	{
		return;
	}

	br->position = position + multiple - position % multiple;
}

static int peek_int_fallback(bit_reader_cxt* br, int bit_count)
{
	int value = 0;
	int byte_index = br->position / 8;
	int bit_index = br->position % 8;
	const unsigned char* buffer = br->buffer;

	while (bit_count > 0)
	{
		if (bit_index >= 8)
		{
			bit_index = 0;
			byte_index++;
		}

		int bits_to_read = bit_count;
		if (bits_to_read > 8 - bit_index)
		{
			bits_to_read = 8 - bit_index;
		}

		const int mask = 0xFF >> bit_index;
		const int current_byte = (mask & buffer[byte_index]) >> (8 - bit_index - bits_to_read);

		value = (value << bits_to_read) | current_byte;
		bit_index += bits_to_read;
		bit_count -= bits_to_read;
	}
	return value;
}